File: Chapter01/scala/packt1/activity/ZipfValidationB.scala
package packt1.solutions

import org.apache.spark.ml.feature.{RegexTokenizer}
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.{DataFrame, SparkSession}
import packt1.HelperScala.{createSession, novellaLocation}

object ZipfValidationB {

  def main(args: Array[String]): Unit = {

    // creating tokenizer as in org.apache.spark.ml.feature.TokenizerTestData.scala
    val tokenizer0 = new RegexTokenizer()
      .setGaps(false)
      .setPattern("\\w+|\\p{Punct}")
      .setInputCol("rawText") // "input" column name
      .setOutputCol("tokens") // "output" column name

    val session: SparkSession = createSession(2, "Zipf Validation")
    import session.implicits._

    val lines: RDD[String] = session.sparkContext.textFile(novellaLocation)
    val linesDf: DataFrame = lines // conversion to DataFrame
      .toDF("rawText")
    val tokenized = tokenizer0 // tokenizer application
      .transform(linesDf)

    tokenized.show(10) // printing a subset

    /*
    +--------------------+--------------------+
    |             rawText|              tokens|
    +--------------------+--------------------+
    |The Project Guten...|[the, project, gu...|
    |                    |                  []|
    |This eBook is for...|[this, ebook, is,...|
    |almost no restric...|[almost, no, rest...|
    |re-use it under t...|[re, -, use, it, ...|
    |with this eBook o...|[with, this, eboo...|
    |                    |                  []|
    |                    |                  []|
    |Title: Heart of D...|[title, :, heart,...|
    |                    |                  []|
    +--------------------+--------------------+
    */
  }

}

File: Chapter01/python/packt1/activity/zipf_validation_b.py
from pyspark.ml.feature import RegexTokenizer
from pyspark.rdd import RDD
from pyspark.sql import SparkSession, DataFrame
from pyspark.sql.types import Row

from Chapter01.python.packt1.helper_python import create_session, novella_location

if __name__ == "__main__":
    session: SparkSession = create_session(2, "Zipf validation")

    reTokenizer = RegexTokenizer(inputCol="val", outputCol="tokens", pattern="\\w+|\\p{Punct}", gaps=False)
    lines: RDD = session.sparkContext.textFile(novella_location)
    row = Row("val")
    df = lines \
        .map(row) \
        .toDF()

    tokenized: DataFrame = reTokenizer.transform(df)
    tokenized.show(10)
    # +--------------------+--------------------+
    # |                 val|              tokens|
    # +--------------------+--------------------+
    # |The Project Guten...|[the, project, gu...|
    # |                    |                  []|
    # |This eBook is for...|[this, ebook, is,...|
    # |almost no restric...|[almost, no, rest...|
    # |re-use it under t...|[re, -, use, it, ...|
    # |with this eBook o...|[with, this, eboo...|
    # |                    |                  []|
    # |                    |                  []|
    # |Title: Heart of D...|[title, :, heart,...|
    # |                    |                  []|
    # +--------------------+--------------------+
